src/python/pants/engine/internals/graph_test.py

```diff
@@ -133,6 +124,83 @@ def get_target(name: str) -> Target:
         assert transitive_targets.dependencies == FrozenOrderedSet([d1, d2, d3, t2, t1])
         assert transitive_targets.closure == FrozenOrderedSet([root, d2, d1, d3, t2, t1])
 
+    def test_transitive_targets_tolerates_subtarget_cycles(self) -> None:
+        """For generated subtargets, we should tolerate cycles between targets.
+
+        This only works with generated subtargets, so we use explicit file dependencies in this
+        test.
+        """
+        self.create_files("", ["dep.txt", "t1.txt", "t2.txt"])
+        self.add_to_build_file(
+            "",
+            dedent(
+                """\
+                target(name='dep', sources=['dep.txt'])
+                target(name='t1', sources=['t1.txt'], dependencies=['dep.txt', 't2.txt'])
+                target(name='t2', sources=['t2.txt'], dependencies=['t1.txt'])
+                """
+            ),
+        )
+        result = self.request_single_product(
+            TransitiveTargets,
+            Params(Addresses([Address.parse("//:t2")]), create_options_bootstrapper()),
+        )
+        assert len(result.roots) == 1
+        assert result.roots[0].address == Address.parse("//:t2")
+        assert [tgt.address for tgt in result.dependencies] == [
+            Address("", target_name="t1.txt", generated_base_target_name="t1"),
+            Address("", target_name="dep.txt", generated_base_target_name="dep"),
+            Address("", target_name="t2.txt", generated_base_target_name="t2"),
+        ]
+
+    def assert_failed_cycle(self, address_str: str, cyclic_address_str: str) -> None:
+        with self.assertRaisesRegex(
+            Exception,
+            f"(?ms)Dependency graph contained a cycle:.*-> {cyclic_address_str}.*-> {cyclic_address_str}.*",
```

Comment:
It'd be great to instead do an exact string match that we are tracing the entire path correctly, rather than only the start and end. This should be easier than before because we no longer have all the engine params, and only have addresses.

Reply:
That makes for a more fragile test, because it is (even more) subject to string formatting. How would you feel about testing using the `CycleException.path`?

Reply:
I would still prefer the exact string test because we are trying to improve the quality of error messages in Pants, and part of that is ensuring that we format them well. 

If we decide to render the error differently, we can decide to update the test.

--

But I won't block on that. It's fine if you want to use `.path`.